packages/backend-auth/src/userpool_access_policy_factory.ts

```diff
@@ -0,0 +1,148 @@
+import { IUserPool } from 'aws-cdk-lib/aws-cognito';
+import { Policy, PolicyStatement } from 'aws-cdk-lib/aws-iam';
+import { Stack } from 'aws-cdk-lib';
+import { AmplifyFault } from '@aws-amplify/platform-core';
+import { UserpoolAction } from './types.js';
+
+export type Permission = {
+  actions: UserpoolAction[];
+  /**
+   * An s3 prefix that defines the scope of the actions
+   */
+  resources: string[];
+};
+
+/**
+ * Generates IAM policies scoped to a single bucket
+ */
+export class UserPoolAccessPolicyFactory {
+  private readonly namePrefix = 'userpoolAccess';
+  private readonly stack: Stack;
+
+  private policyCount = 1;
+
+  /**
+   * Instantiate with the bucket to generate policies for
+   */
+  constructor(private readonly userpool: IUserPool) {
+    this.stack = Stack.of(userpool);
+  }
+
+  createPolicy = (actions: Readonly<Set<UserpoolAction>>) => {
+    if (actions.size === 0) {
+      throw new AmplifyFault('EmptyPolicyFault', {
+        message: 'At least one action must be specified',
+      });
+    }
+
+    const statements: PolicyStatement[] = [];
+
+    actions.forEach((action) => {
+      statements.push(this.getStatement(action));
+    });
+
+    return new Policy(this.stack, `${this.namePrefix}${this.policyCount++}`, {
+      statements,
+    });
+  };
+
+  private getStatement = (action: UserpoolAction) =>
+    new PolicyStatement({
+      actions: actionPolicyMap[action],
+      resources: [this.userpool.userPoolArn],
+    });
+}
+
+const actionPolicyMap: Record<UserpoolAction, string[]> = {
```

Comment:
Aren't we essentially breaking the "Principle of Least Privilege" here by combining these multiple permissions against generic umbrella terms `create`, `list` etc?

The actions here should be very specifically named that maps to the strict set of permissions required for those actions.

Reply:
I think it's fine for a single "action" to map to a handful of different IAM actions. for example maybe "creating users" requires AdminCreateUser, ConfirmSignUp and SetUserMFAPreference (just making that up). It would be okay to combine that all together under one "create user" use case (along with documentation on exactly what IAM actions will be enabled). But this huge list of actions that doesn't seem to have high cohesion is definitely not what we want.

Reply:
Yes combining the IAM actions is fine, but not under a generic action such as `create` or `read`. This is too ambiguous for resource that has multiple things to create/read etc.